### PR TITLE
Handle more logfmt serialization corner cases and errors

### PR DESCRIPTION
**Pull request by bloodearnest**

### CODE REVIEW

File: tests/test_sentry.py [hunk 1]
```diff
@@ -164,6 +164,7 @@ def app(environ, sr):
 
     mw = talisker.sentry.get_middleware(app)
     context = mw.client.context
+    context.clear()
```
Comment (bloodearnest) [inline tests/test_sentry.py]: Note: this is related to getting the tests to pass when you have SENTRY_DSN set in the environment, seemed reasonable to include it.

File: talisker/logs.py [hunk 2]
```diff
@@ -359,36 +365,63 @@ def logfmt_key(self, k):
...
+            v = '"' + str(type(v)) + '"'
 
         return v
 
+    def safe_string(self, s, max, truncate):
+        s = s.strip()
+
+        # no new lines allowed, so truncate at the first new line
+        if '\n' in s:
+            s = s.split('\n', 1)[0] + truncate
```
Comment (sparkiegeek) [inline talisker/logs.py]: hmm, if the first line is longer than the maximum length then we'll add truncate twice!

**Comment by bloodearnest** [comment]
Hah, good point.

File: talisker/logs.py [hunk 3]
```diff
@@ -341,54 +349,100 @@ def logfmt_atoms(self, structured):
...
-        elif isinstance(v, dict):
-            return '"' + self.TRUNCATED + '"'
+        elif isinstance(v, numbers.Number):
+            v = str(v)
+        else:
+            v = '"' + str(type(v)) + '"'
 
         return v
 
+    def safe_string(self, s, max, truncate_str=None):
```
Comment (sparkiegeek) [inline talisker/logs.py]: why is `truncate_str` optional? you're passing it through for both the key and value cases..

File: talisker/logs.py [hunk 4]
```diff
@@ -341,54 +349,100 @@ def logfmt_atoms(self, structured):
...
+            truncated = True
+
+        # maximum size, to prevent overloading aggregator (accidental or
+        # malicious)
+        if len(s) > max:
+            s = s[:max]
+            truncated = True
+
+        # logstash kv filter cannot handle escaped ", so we just strip
+        s = s.replace('"', '')
```
Comment (sparkiegeek) [inline talisker/logs.py]: there's a minor bug-let here where a string that is composed of one or more `"` and is above max-size could in fact be unnecessarily truncated (since we remove the `"` characters after testing for length)

File: talisker/logs.py (hunk 3 above)
Comment (bloodearnest) [reply talisker/logs.py]: Yes, and artifact for a previous iteration, will remove

File: talisker/logs.py (hunk 4 above)
Comment (bloodearnest) [reply talisker/logs.py]: hmm, right. I'm not overly concerned with that. The truncate length is an arbitrary limit to stop unbounded message size. A few chars +/- doesn't worry me. And doing it after truncate makes it faster, avoids needless replacements for things that are going to be discarded anyway.

Comment (sparkiegeek) [reply talisker/logs.py]: :+1:
